File: numerical_analysis/example_code/topic1/duffing.py
# Duffing Oscillator

from math import cos

# physics parameters
omega = 2.4
gamma = 0.1
a = 0.5
b = 0.25
F0 = 2.0

# Duffing equation
def dxvdt(t, xv):
    dxdt = xv[1]
    dvdt = - gamma * xv[1] + 2 * a * xv[0] - 4 * b * xv[0]**3
    dvdt += F0 * cos(omega * t)
    return [dxdt, dvdt]

# get input from user and initialize
print(' Duffing Oscillator Simulation')
print(' =============================')
x = float(input(' Enter initial x: '))
v = float(input(' Enter initial v: '))
xv = [x, v]
dt = float(input(' Enter time step dt: '))
t_max = float(input(' Enter integration time: '))
file_name = input(' Enter output file name: ')
file = open(file_name, "w")
t = 0.0
file.write(str(t) + '\t' + str(xv[0]) + '\t' + str(xv[1]) + '\n')

# main integration loop
while t < t_max:
    # 4th order Runge-Kutta step
    k1 = [ dt * d for d in dxvdt(t, xv) ]
    xv_step = [ xv[i] + k1[i] / 2 for i in range(2) ]
    k2 = [ dt *  d for d in dxvdt(t + 0.5 * dt, xv_step) ]
    xv_step = [ xv[i] + k2[i] / 2 for i in range(2) ]
    k3 = [ dt * d for d in dxvdt(t + 0.5 * dt, xv_step) ]
    xv_step = [ xv[i] + k3[i] for i in range(2) ]
    k4 = [ dt * d for d in dxvdt(t + dt, xv_step) ]
    for i in range(2):
        xv[i] += (k1[i] + 2 * k2[i] + 2 * k3[i] + k4[i]) / 6.0
    t += dt
    file.write(str(t) + '\t' + str(xv[0]) + '\t' + str(xv[1]) + '\n')

file.close()
print(' Output in file:', file_name)

File: numerical_analysis/example_code/topic1/duffing.f90
! Duffing Oscillator

program duffing
    implicit none
! real kind with decimal precision() >= 14 and exponent range() >= 300
    integer, parameter :: dp = selected_real_kind(14, 300)
! constants
    real(dp), parameter :: zero = 0._dp, one = 1._dp, two = 2._dp
    real(dp), parameter :: half = one / two, six = (one + two) * two
! physics parameters
    real(dp) :: omega = 2.4_dp, gamma = 0.1_dp
    real(dp) :: a = 0.5_dp, b = 0.25_dp
    real(dp) :: F0 = 2.0_dp
! variables
    real(dp), dimension(2) :: xv, dxvdt
    real(dp) :: t, t_max, dt
    real(dp), dimension(2) :: k1, k2, k3, k4, xv_step
    character(len=50) :: file_name
! get input from user and initialize
    print *, 'Duffing Oscillator Simulation'
    print *, '============================='
    print *, 'Enter initial x and v: '
    read *, xv(1), xv(2)
    print *, 'Enter time step dt'
    read *, dt
    print *, 'Enter integration time: '
    read *, t_max
    print *, 'Enter output file name: '
    read *, file_name
    open(unit=2, file=file_name)
    t = zero
    write(2, *) t, xv(1), xv(2)
! main integration loop
    do
        if (t >= t_max) exit
!    4th order Runge-Kutta step
        call find_dxvdt(t, xv, dxvdt)
        k1 = dt * dxvdt
        xv_step = xv + k1 / two
        call find_dxvdt(t + half * dt, xv_step, dxvdt)
        k2 = dt * dxvdt
        xv_step = xv + k2 / two
        call find_dxvdt(t + half * dt, xv_step, dxvdt)
        k3 = dt * dxvdt
        xv_step = xv + k3
        call find_dxvdt(t + dt, xv_step, dxvdt)
        k4 = dt * dxvdt
        xv = xv + (k1 + two * k2 + two * k3 + k4) / six
        t = t + dt
        write(2, *) t, xv(1), xv(2)
    end do
    print *, 'Output in file: ', file_name
    close(2)

! Duffing equation
contains
    subroutine find_dxvdt(t, xv, dxvdt)
        implicit none
        real(dp), intent(in) :: t
        real(dp), dimension(2), intent(in) :: xv
        real(dp), dimension(2), intent(out) :: dxvdt
        dxvdt(1) = xv(2)
        dxvdt(2) = - gamma * xv(2) + 2 * a * xv(1) - 4 * b * xv(1)**3
        dxvdt(2) = dxvdt(2) + F0 * cos(omega * t)
    end subroutine find_dxvdt
end program duffing
